chi/widgets.py

```diff
@@ -175,125 +168,88 @@ def update_selected_node(node_type):
         return new_nodes
 
-    if gpu_count is None:
-        pass
-    elif type(gpu_count) is int and gpu_count > 0:
-        if gpu is False:
-            raise IllegalArgumentError(f"input 'gpu' not compatible with "
-                                       f"input 'gpu_count': False and"
-                                       f" {gpu_count} contradict each other")
-
-        avail_nodes = _find_gpu_helper(avail_nodes, gpu_count)
-    elif type(gpu_count) is int and gpu_count == 0:
-        if gpu is True:
-            raise IllegalArgumentError(f"input 'gpu' not compatible with "
-                                       f"input 'gpu_count': True and"
-                                       f" 0 contradict one another")
-        gpu = False
-    else:
-        raise IllegalArgumentError(f"parameter 'gpu_count' expected type 'int'"
-                                   f" > 0, received input '{gpu_count}' of "
-                                   f"type '{type(gpu_count).__name__}'")
-
-    if gpu is None:
-        pass
-    elif gpu is True and gpu_count is None:
-        avail_nodes = _find_gpu_helper(avail_nodes)
-    elif gpu is False:
-        avail_nodes = _find_gpu_helper(avail_nodes, 0)
-    else:
-        raise IllegalArgumentError(f"parameter 'gpu' expected type 'bool'"
-                                   f", received type "
-                                   f"'{type(gpu).__name__}'")
-
-    if storage_size_gb is None:
-        pass
-    elif type(storage_size_gb) is int and storage_size_gb >= 0:
-        avail_nodes = _minimum_storage_helper(avail_nodes, storage_size_gb)
-    else:
-        raise IllegalArgumentError(f"parameter 'storage_size_gb' expected type"
-                                   f" 'int' > 0, received input "
-                                   f"'{storage_size_gb}' of type "
-                                   f"'{type(storage_size_gb).__name__}'")
-
-    if architecture is None:
-        pass
-    elif type(architecture) is str:
-        avail_nodes = _find_architecture_helper(avail_nodes, architecture)
-    else:
-        raise IllegalArgumentError(f"parameter 'architecture' expected type"
-                                   f" 'str', received type"
-                                   f" '{type(architecture).__name__}'")
-
-    if ssd is None:
-        pass
-    elif type(ssd) is bool:
-        avail_nodes = _find_ssd_helper(avail_nodes, ssd)
-    else:
-        raise IllegalArgumentError(f"parameter 'ssd' expected type"
-                                   f" 'bool', received type"
-                                   f" '{type(ssd).__name__}'")
-
+    if has_gpu is not None:
+        avail_nodes = has_gpu_helper(avail_nodes, has_gpu)
+    if gpu_count is not None:
+        if has_gpu is True and gpu_count == 0:
+            raise ValueError("'has_gpu' specified, but requested nodes "
+                             "with 0 GPUs")
+        if has_gpu is False and gpu_count > 0:
```

Comment:
```suggestion
        if not has_gpu and gpu_count is not None:
```
We probably want to catch if the user has specified `gpu_count` at all. If they have, it was probably a mistake. This comparison operator will also throw this if `gpu_count` is `None`:

```
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

Reply:
I agree about everything except the `TypeError`; this `if` clause is nested in `if gpu_count is not None` so there wouldn't be an error. If we added `gpu_count is not None` we'd end up with:
```
    if gpu_count is not None:
        if has_gpu and gpu_count <= 0:
            raise ValueError("'has_gpu' specified, but requested nodes "
                             "with 0 GPUs")
        if not has_gpu and gpu_count is not None:
            raise ValueError("'gpu_count' specified, but requested nodes "
                             "without GPU")
```